stdlib/int.mli

```diff
@@ -0,0 +1,140 @@
+
+external succ : int -> int = "%succint"
+(** [succ x] is [add x 1]. *)
+
+external pred : int -> int = "%predint"
+(** [pred x] is [sub x 1]. *)
+
+val abs : int -> int
+(** [abs x] is the absolute value of [x]. That is [x] if [x] is positive
+    and [neg x] if [x] is negative. {b Warning.} This may be negative if
+    the argument is {!min_int}. *)
+
+val max_int : int
+(** [max_int] is the greatest representable integer,
+    [2{^[Sys.int_size - 1]} - 1]. *)
+
+val min_int : int
+(** [min_int] is the smallest representable integer,
+    [-2{^[Sys.int_size - 1]}]. *)
+
+external logand : int -> int -> int = "%andint"
+(** [logand x y] is the bitwise logical and of [x] and [y]. *)
+
+external logor : int -> int -> int = "%orint"
+(** [logor x y] is the bitwise logical or of [x] and [y]. *)
+
+external logxor : int -> int -> int = "%xorint"
+(** [logxor x y] is the bitwise logical exclusive or of [x] and [y]. *)
+
+val lognot : int -> int
+(** [lognot x] is the bitwise logical negation of [x]. *)
+
+external shift_left : int -> int -> int = "%lslint"
+(** [shift_left x n] shifts [x] to the left by [n] bits. The result
+    is unspecified if [n < 0] or [n > ]{!Sys.int_size}. *)
+
+external shift_right : int -> int -> int = "%asrint"
+(** [shift_right x n] shifts [x] to the right by [n] bits. This is an
+    arithmetic shift: the sign bit of [x] is replicated and inserted
+    in the vacated bits. The result is unspecified if [n < 0] or
+    [n > ]{!Sys.int_size}. *)
+
+external shift_right_logical : int -> int -> int = "%lsrint"
+(** [shift_right x n] shifts [x] to the right by [n] bits. This is a
+    logical shift: zeroes are inserted in the vacated bits regardless
+    of the sign of [x]. The result is unspecified if [n < 0] or
+    [n > ]{!Sys.int_size}. *)
+
+(** {1:preds Predicates and comparisons} *)
+
+val equal : int -> int -> bool
+(** [equal x y] is [true] iff [x = y]. *)
+
+val compare : int -> int -> int
+(** [compare x y] is {!Stdlib.compare}[ x y] but more efficient. *)
+
+(** {1:convert Converting} *)
+
+external to_float : int -> float = "%floatofint"
+(** [to_float x] is [x] as a floating point number. *)
+
+external of_float : float -> int = "%intoffloat"
+(** [of_float x] truncates [x] to an integer. The result is
+    unspecified if the argument is [nan] or falls outside the range of
+    representable integers. *)
+
+val of_string : string -> int option
+(** [of_string s] is [Some s] if [s] can be parsed to an integer
+    in the range representable by the type [int] (note that this
+    depends on {!Sys.int_size}) and [None] otherwise.
+
+    The string is read in decimal (by default, or if the string
+    begins with [0u]), in hexadecimal (if it begins with [0x] or
+    [0X]), in octal (if it begins with [0o] or [0O]), or in binary
+    (if it begins with [0b] or [0B]).
+
+    The [0u] prefix reads the input as an unsigned integer in the range
```

Comment:
Some tweaks to the spec:

- The base prefixes come after the sign (- / +) (i.e. "beings with" is not accurate).
- 0U is also supported.
- The underscore cannot occur "in the prefix" (before the sign, between the sign and the base, in the middle of the base).
- 0u prefix allow negative numbers as well (upto something like -2*max_int-1).

Also, a design question: do you know why we cannot specify "unsigned" for non-decimal bases?

Reply:
Done.